test/runner/createTestServer.js

```diff
@@ -1,70 +1,186 @@
+					"<script src=\"/test/runner/listeners.js\"></script></head>"
+				)
+			);
+		} else {
+			next();
+		}
 	} );
 
 	// Bind the reporter
-	app.post(
-		"/api/report",
-		bodyParser.json( { limit: "50mb" } ),
-		async( req, res ) => {
-			if ( report ) {
-				const response = await report( req.body );
-				if ( response ) {
-					res.json( response );
-					return;
-				}
+	use( async( req, res, next ) => {
+		if ( req.url !== "/api/report" || req.method !== "POST" ) {
+			return next();
+		}
+		let body;
+		try {
+			body = JSON.parse( await getRawBody( req ) );
+		} catch ( error ) {
+			if ( error.code === "ECONNABORTED" ) {
+				return;
 			}
-			res.sendStatus( 204 );
+			console.error( error );
+			res.writeHead( 400, { "Content-Type": "application/json" } );
+			res.end( JSON.stringify( { error: "Invalid JSON" } ) );
+			return;
 		}
-	);
-
-	// Handle errors from the body parser
-	app.use( bodyParserErrorHandler() );
+		const response = await report( body );
+		if ( response ) {
+			res.writeHead( 200, { "Content-Type": "application/json" } );
+			res.end( JSON.stringify( response ) );
+		} else {
+			res.writeHead( 204 );
+			res.end();
+		}
+	} );
 
 	// Hook up mock server
-	app.use( mockServer() );
+	use( mockServer() );
 
 	// Serve static files
-	app.post( "/test/data/name.html", ( _req, res ) => {
-		res.send( nameHTML );
-	} );
+	const validMimeTypes = {
+		".js": "application/javascript",
+		".css": "text/css",
+		".html": "text/html",
+		".xml": "application/xml",
+		".xhtml": "application/xhtml+xml",
+		".jpg": "image/jpeg",
+		".png": "image/png",
+		".svg": "image/svg+xml",
+		".ico": "image/x-icon",
+		".map": "application/json",
+		".txt": "text/plain",
+		".log": "text/plain"
+	};
+	use( async( req, res, next ) => {
+		if (
+			req.url.startsWith( "/dist/" ) ||
+			req.url.startsWith( "/src/" ) ||
+			req.url.startsWith( "/test/" ) ||
+			req.url.startsWith( "/external/" )
+		) {
+			const file = req.parsedUrl.pathname.slice( 1 );
+			const ext = file.slice( file.lastIndexOf( "." ) );
 
-	app.use( "/dist", express.static( "dist" ) );
-	app.use( "/src", express.static( "src" ) );
-	app.use( "/test", express.static( "test" ) );
-	app.use( "/external", express.static( "external" ) );
+			// Allow POST to .html files in tests
+			if (
+				req.method !== "GET" &&
+				req.method !== "HEAD" &&
+				( ext !== ".html" || req.method !== "POST" )
+			) {
+				return next();
+			}
+			const mimeType = validMimeTypes[ ext ];
+			if ( mimeType ) {
+				try {
+					await stat( file );
+				} catch ( error ) {
+					res.writeHead( 404 );
+					res.end();
+					return;
+				}
+				res.writeHead( 200, { "Content-Type": mimeType } );
+				createReadStream( file ).pipe( res ).on( "error", ( error ) => {
+					console.error( error );
+					res.writeHead( 500 );
+					res.end();
+				} );
+			} else {
+				console.error( `Invalid file extension: ${ ext }` );
+				res.writeHead( 404 );
+				res.end();
+			}
+		} else {
+			next();
```

Comment:
This `if` above is pretty big; can we short-circuit the `next()` call here by reversing the initial `if`? I think you've already done it for some other bigger route handlers.